Project architecture/MVVM.md

```diff
@@ -110,28 +152,68 @@ open class BaseViewModel<State : Any, Event : Any> : ViewModel() {
 
 When we have a BaseViewModel implementation, for example LoginViewModel, we have to provide it to the View. This can be achieved using a utility method from the ViewModelProviders class:
 
-```
+```kotlin
 ViewModelProviders.of(this).get(LoginViewModel::class.java)
 
 ```
 
 In the above code, `this` represents the LifecycleOwner. In our case, as mentioned earlier, this is either an Activity or a Fragment. The above code returns a ViewModel instance which we use to observe LiveData objects.
 
-```
-loginViewModel.viewStateData().observe(this, { state ->
-	// Update the UI state
-})
+```kotlin
+loginViewModel.viewStateData().observe(this) { state ->
+    // Update the UI state
+}
 
-loginViewModel.viewEventData().observe(this, { event ->
-	// Handle event
-})
+loginViewModel.viewEventData().observe(this) { event ->
+    // Handle event
+}
 
 ```
 
 This is now enough to establish a connection between the View and ViewModel.
 
 So far, we've only concentrated on the View and ViewModel, but what about the Model in MVVM? The Model is responsible for managing the data received from a specific data source (Database, Network, etc.), completely UI independent. This means that the Model should expose its data only to the ViewModel. The ViewModel can also request some data from the Model, so it is a two-way communication.
 
+### Implementation using StateFlow and SharedFlow
+
+If we want to use StateFlow and SharedFlow as a substitution to the LiveData objects mentioned above, our BaseViewModel will look something like this:
+
+```kotlin
+abstract class BaseViewModel<State : Any, Event : Any>(private val initialState: State) : ViewModel() {
+
+    val state: StateFlow<State> get() = stateFlow
+    val commonState: StateFlow<CommonState> get() = commonStateFlow
+    val event: SharedFlow<Event> get() = eventFlow
+
+    private val stateFlow: MutableStateFlow<State> by lazy { MutableStateFlow(initialState) }
+    private val commonStateFlow = MutableStateFlow(CommonState())
+    private val eventFlow: MutableSharedFlow<Event> = MutableSharedFlow()
+}
```

Comment:
Again, this should go into a more extensive discussion about the restructuring handbook. I think we should add chapter about the common testing setup for architecture (base test classes, how to test livedata, flows etc)
I remember recently there was some question about it in #team-android